generator/src/Google/Service/Generator/ServiceGenerator.php

```diff
@@ -0,0 +1,119 @@
+<?php
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Service\Generator;
+
+class ServiceGenerator {
+  const PREFERRED = [
+    'admin:datatransfer_v1' => true,
+    'admin:directory_v1' => true,
+    'discovery:v1' => false,
+  ];
+  public $path = __DIR__ . "/../../../../../src/Google/Service";
+
+  function __construct($path = '') {
+    if ($path != '') {
+      $this->path = $path;
+    }
+  }
+
+  function generate($discoveryURL) {
+    $old_error_handler = set_error_handler([$this, 'errorHandler']);
+
+    $service = new Service(json_decode(file_get_contents($discoveryURL), 1));
+    $path = "$this->path/$service->canonicalName";
+
+    if (!is_dir("$path/Resource")) {
+      mkdir("$path/Resource", 0755, true);
+    }
+
+    $smarty = new \Smarty;
+    $smarty->setTemplateDir([__DIR__. '/../../../../templates']);
+    $smarty->setCompileDir(sys_get_temp_dir().'/gs_tpls');
+
+    $smarty->assign("Service", $service);
+    $smarty->assign("CopyrightYear", 2014);
+    file_put_contents("$path.php", $smarty->fetch('service.tpl'));
+
+    foreach ($service->getAllResources() as $resource) {
+      $smarty->assign("Resource", $resource);
+      $filename = "$path/Resource/" . $resource->getClassName() . ".php";
+      file_put_contents($filename, $smarty->fetch('resource.tpl'));
+    }
+
+    foreach ($service->getSchemas() as $schema) {
+      $classname = $schema->getName();
+      $smarty->assign("ClassName", $classname);
+
+      if (!$schema->properties) {
+        // error_log("$path/$classname.php is blank!");
+        if ($classname == "JsonObject") { //TODO
```

Comment:
Re: TODO

Should we finish this TODO before the merge?

Reply:
Done.